generic-sensor/generic-sensor-tests.js

```diff
@@ -0,0 +1,158 @@
+function runGenericSensorTests(sensorType, verifyReading) {
+  async_test(t => {
+    let sensor = new sensorType();
+    sensor.onchange = t.step_func_done(() => {
+      assert_true(verifyReading(sensor));
+      assert_equals(sensor.state, "activated");
+      sensor.stop();
+    });
+    sensor.onerror = t.step_func_done(event => {
+      assert_unreached(event.error.name + ":" + event.error.message);
+    });
+    sensor.start();
+  }, "event change fired");
+
+  async_test(t => {
+    let sensor1 = new sensorType();
+    let sensor2 = new sensorType();
+    sensor1.onactivate = t.step_func_done(() => {
+      // Reading values are correct for both sensors.
+      assert_true(verifyReading(sensor1));
+      assert_true(verifyReading(sensor2));
+
+      //After first sensor stops its reading values are null,
+      //reading values for the second sensor remains
+      sensor1.stop();
+      assert_true(verifyReading(sensor1, true /*should be null*/));
+      assert_true(verifyReading(sensor2));
+      sensor2.stop();
+      assert_true(verifyReading(sensor2, true /*should be null*/));
+    });
+    sensor1.onerror = t.step_func_done(event => {
+      assert_unreached(event.error.name + ":" + event.error.message);
+    });
+    sensor2.onerror = t.step_func_done(event => {
+      assert_unreached(event.error.name + ":" + event.error.message);
+    });
+    sensor1.start();
+    sensor2.start();
+  }, "sensor reading is correct");
+
+  async_test(t => {
+    let sensor = new sensorType();
+    let cachedTimeStamp1;
+    sensor.onactivate = () => {
+      cachedTimeStamp1 = sensor.timestamp;
+    };
+    sensor.onerror = t.step_func_done(event => {
+      assert_unreached(event.error.name + ":" + event.error.message);
+    });
+    sensor.start();
+    t.step_timeout(() => {
+      sensor.onchange = t.step_func_done(() => {
+        //sensor.timestamp changes.
+        let cachedTimeStamp2 = sensor.timestamp;
+        assert_greater_than(cachedTimeStamp2, cachedTimeStamp1);
+        sensor.stop();
+        t.done();
+      });
+    }, 1000);
+  }, "timestamp is updated when time passes");
+
+  test(() => {
+    let sensor, start_return;
+    sensor = new sensorType();
+    sensor.onerror = () => {
+      assert_unreached(event.error.name + ":" + event.error.message);
+    };
+    //The default sensor.state is 'unconnected'
+    assert_equals(sensor.state, "unconnected");
+    start_return = sensor.start();
+    //The sensor.state changes to 'activating' after sensor.start()
+    assert_equals(sensor.state, "activating");
+    //TODO: The permission is not ready.
+    //the sensor.start() return undefined
+    assert_equals(start_return, undefined);
+    sensor.stop();
+  }, "sensor.start() is correct");
```

Comment:
That's vague. You might want to add more specific tests instead. e.g. "`sensor.start` doesn't throw in [STATE] state," "`sensor.start` returns undefined," "`sensor.start` triggers activation," etc.